### ISSUE TITLE
Possible E0301 false positive

### ISSUE THREAD

**Issue by Nadrieril** [rust-lang/rust#59803]
This code fails with `error[E0301]: cannot mutably borrow in a pattern guard`:
```rust
match Some(1) {
    x if x.iter().all(|x| *x == 1) => {}
    _ => {}
}
```

<details><summary>Error</summary>

```
error[E0301]: cannot mutably borrow in a pattern guard
   --> blah/src/lib.rs:230:14
    |
230 |         x if x.iter().all(|x| *x == 1) => {}
    |              ^^^^^^^^ borrowed mutably in pattern guard
    |
    = help: add #![feature(bind_by_move_pattern_guards)] to the crate attributes to enable
```
</details>

However the following works:
```rust
let is_all_1 = |x: &Option<usize>| x.iter().all(|x| *x == 1);
match Some(1) {
    x if is_all_1(&x) => {}
    _ => {}
}
```
Which is rather surprising and looks like a false positive to me. Could the compiler allow a match if the lifetime of the mutable borrow is limited to the match guard ?

**Comment by Nadrieril** [comment]
This appears to be fixed [playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=11c2f7aa1a9d274198702f2aeb0fee57)
